### PR TITLE
feat: add telemetry/alerts via sns

### PR DESCRIPTION
**Pull request by iamKunalGupta**

### CODE REVIEW

File: flow/shared/alerting/alerting.go [hunk 1]
```diff
@@ -213,4 +255,7 @@ func (a *Alerter) LogFlowInfo(ctx context.Context, flowName string, info string)
 		logger.LoggerFromCtx(ctx).Warn("failed to insert flow info", slog.Any("error", err))
 		return
 	}
+	// TODO Maybe too much noise here
+	//a.sendTelemetryMessage(ctx, flowName, info, INFO)
```
Comment (serprex) [inline flow/shared/alerting/alerting.go:259]:
```suggestion
```
Agreed

**Review by serprex — APPROVED** [review]

File: flow/shared/alerting/alerting.go [hunk 2]
```diff
@@ -194,6 +208,28 @@ func (a *Alerter) checkAndAddAlertToCatalog(ctx context.Context, alertKey string
...
 }
 
+func (a *Alerter) sendTelemetryMessage(ctx context.Context, flowName string, more any, level telemetry.Level) {
+	if a.telemetrySender != nil {
+		deployUuidPrefix := ""
+		deployUuid := peerdbenv.PeerDBDeploymentUID()
+		if deployUuid != "" {
+			deployUuidPrefix = fmt.Sprintf("[%s] ", deployUuid)
+		}
+
```
Comment (iskakaushik) [inline flow/shared/alerting/alerting.go:218]: we can remove this as the SendMessage itself has `DeploymentUID`, we use this to identify a deployment.

File: flow/shared/alerting/alerting.go [hunk 3]
```diff
@@ -55,9 +58,20 @@ func NewAlerter(catalogPool *pgxpool.Pool) (*Alerter, error) {
 	if catalogPool == nil {
 		return nil, errors.New("catalog pool is nil for Alerter")
 	}
-
+	snsTopic := os.Getenv("TELEMETRY_AWS_SNS_TOPIC_ARN")
+	var snsMessageSender telemetry.Sender
+	if snsTopic != "" {
+		var err error
+		snsMessageSender, err = telemetry.NewSNSMessageSenderWithNewClient(context.TODO(), &telemetry.SNSMessageSenderConfig{
```
Comment (iskakaushik) [inline flow/shared/alerting/alerting.go:65]: Is this the right context to use over here? Why this vs `context.Background` ?

File: flow/shared/alerting/alerting.go [hunk 4]
```diff
@@ -55,9 +58,20 @@ func NewAlerter(catalogPool *pgxpool.Pool) (*Alerter, error) {
 	if catalogPool == nil {
 		return nil, errors.New("catalog pool is nil for Alerter")
 	}
-
+	snsTopic := os.Getenv("TELEMETRY_AWS_SNS_TOPIC_ARN")
+	var snsMessageSender telemetry.Sender
+	if snsTopic != "" {
```
Comment (iskakaushik) [inline flow/shared/alerting/alerting.go:63]: can we log something like - "successfully registered telemetry client" or something.

File: flow/shared/alerting/alerting.go [hunk 5]
```diff
@@ -203,6 +239,12 @@ func (a *Alerter) LogFlowError(ctx context.Context, flowName string, err error)
 		logger.LoggerFromCtx(ctx).Warn("failed to insert flow error", slog.Any("error", err))
 		return
 	}
+	a.sendTelemetryMessage(ctx, flowName, err, telemetry.ERROR)
+}
+
+func (a *Alerter) LogFlowStart(ctx context.Context, flowName string, info string) {
```
Comment (iskakaushik) [inline flow/shared/alerting/alerting.go:245]: can we add one for when snapshot finishes as well?

File: flow/shared/alerting/alerting.go (hunk 3 above)
Comment (iamKunalGupta) [reply flow/shared/alerting/alerting.go:65]: was not sure which context to pull as the method did not accept one yet. can shift to `context.Background`

Comment (serprex) [reply flow/shared/alerting/alerting.go:65]: Function should be updated to take a context instead

**Review by heavycrystal — APPROVED** [review]

File: flow/shared/telemetry/sns_message_sender.go [hunk 6]
```diff
@@ -58,7 +58,9 @@ func (s *SNSMessageSenderImpl) SendMessage(ctx context.Context, subject string,
 }
 
 func NewSNSMessageSenderWithNewClient(ctx context.Context, config *SNSMessageSenderConfig) (SNSMessageSender, error) {
-	client, err := newSnsClient(ctx)
+	// Topic Region must match client region
+	region := strings.Split(strings.TrimPrefix(config.Topic, "arn:aws:sns:"), ":")[0]
```
Comment (serprex) [inline flow/shared/telemetry/sns_message_sender.go:62]: Use strings.Cut: https://pkg.go.dev/strings#Cut

File: flow/shared/telemetry/sns_message_sender.go [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+	})
+	if err != nil {
+		return nil, err
+	}
+	return publish.MessageId, nil
+}
+
+func NewSNSMessageSenderWithNewClient(ctx context.Context, config *SNSMessageSenderConfig) (SNSMessageSender, error) {
+	// Topic Region must match client region
+	region := strings.Split(strings.TrimPrefix(config.Topic, "arn:aws:sns:"), ":")[0]
```
Comment (serprex) [inline flow/shared/telemetry/sns_message_sender.go:62]: Use `strings.Cut`: https://pkg.go.dev/strings#Cut
May also want `strings.CutPrefix` with error if Topic is lacking prefix

**Pull request merged**